lib/utils/diff.ex

```diff
@@ -0,0 +1,94 @@
+defmodule LambdaEthereumConsensus.Utils.Diff do
+  @moduledoc """
+  A module to compare between different structures and get their differences.
+  Useful for debugging.
+
+  Most typical cases are maps and lists, which recurse in their elements.
+  Any other type is just compared with ==. Tuples may be added if needed,
+  but it wasn't the case yet.
+
+  Possible results:
+  - When comparing any two elements that are equal, the result is :unchanged.
+  - When comparing different maps or lists, the result will be a map containing the keys
+    :added_right, :added_left and :changed.
+
+  Diff attributes:
+  - :added_right or :added_left are lists of elements. When comparing lists, this will
+    just be the elements that were added. In the case of maps, they are keyword lists
+    showing the key and the value that were added.
+  - :changed is always a keyword list. In the case of lists, keys are indices of the
+    changed elements and values are the elements itselfs. In the case of maps, the keys
+    are the keys of the map.
+  """
+  @type structured_diff :: %{
+          optional(:added_left) => [any()],
+          optional(:added_right) => [any()],
+          optional(:changed) => Keyword.t()
+        }
+  @type base_diff :: %{optional(:left) => any(), optional(:right) => any()}
+  @type t :: :unchanged | base_diff() | base_diff()
```

Comment:
The second `base_diff` should be `structured_diff`, right?

Reply:
indeed